emma/OntoEmma.py

```diff
@@ -533,17 +584,208 @@ def _align_string_equiv(s_kb, t_kb):
+
+    @staticmethod
+    def _apply_modh_alignment_strategy(sim_scores, s_kb, t_kb):
+        """
+        Use MODH alignment strategy: use modified hungarian algorithm to determine best global alignment
+        :param sim_scores
+        :param s_kb
+        :param t_kb
+        :return:
+        """
+        alignment = []
+
+        s_len = len(s_kb.entities)
+        t_len = len(t_kb.entities)
+
+        # create and populate graph adjacency matrix
+        cost_mat = np.ones((s_len, t_len))
+
+        for (s_ent_id, t_ent_id), score in sim_scores.items():
+            s_ind = s_kb.get_entity_index(s_ent_id)
+            t_ind = t_kb.get_entity_index(t_ent_id)
+            cost_mat[s_ind][t_ind] = 1.0 - score
+
+        print("Computing best alignment...")
+        hmod = ModifiedHungarian(cost_mat)
+        indices = hmod.compute()
+        print("Length of assignment: %i" % len(indices))
+
+        for row, column in indices:
+            row_ent_id = s_kb.entities[row].research_entity_id
+            col_ent_id = t_kb.entities[column].research_entity_id
+            score = sim_scores[(row_ent_id, col_ent_id)]
+            if score >= constants.SIM_SCORE_THRESHOLD:
+                alignment.append((s_kb.entities[row].research_entity_id,
+                                  t_kb.entities[column].research_entity_id,
+                                  score))
+        return alignment
+
+    def _compute_alignment(self, align_strat, l_scores, n_scores, s_kb, t_kb):
+        """
+        Compute best bipartite alignment between s_kb and t_kb based on scores
+        :param l_scores: entity similarity scores
+        :param n_scores: entity similarity scores updated based on similarity of neighborhood entities
+        :param s_kb:
+        :param t_kb:
+        :return:
+        """
+        if align_strat == "best":
+            return self._apply_best_alignment_strategy(n_scores)
+        elif align_strat == "all":
+            return self._apply_all_alignment_strategy(n_scores)
+        elif align_strat == "modh":
+            return self._apply_modh_alignment_strategy(n_scores, s_kb, t_kb)
+        else:
+            raise NotImplementedError("Error: Unknown alignment strategy.")
+
+    def _compute_neighborhood_similarities(self, scores, s_kb, t_kb, iterations=0):
+        """
+        Compute neighborhood similarities of each pair of entities based on entity similarity scores
+        :param scores: Entity similarity scores
+        :param s_kb: Source KB
+        :param t_kb: Target KB
+        :param iterations: number of iterations for propagating neighborhood similarity
+        :return:
+        """
+        updated_neighborhood_sim = copy(scores)
+
+        # iteratively calculate global similarity scores
+        for i in range(0, iterations):
+            neighborhood_sim = dict()
+
+            # iterate through all alignments
+            for (s_ent_id, t_ent_id), score in tqdm.tqdm(updated_neighborhood_sim.items(),
```

Comment:
How slow is this? Could you add an in-line comment noting that this block of code can be reimplemented more efficiently using a dynamic programming algorithm to avoid processing the same pair multiple times when they show up in multiple neighborhoods.